src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodTranslator.cs

```diff
@@ -73,20 +85,32 @@ public SqlExpression Translate(SqlExpression instance, MethodInfo method, IReadO
             // since non-PG SQL does not support arrays. If the list is a constant we leave it for regular IN
             // (functionality the same but more familiar).
 
-            // TODO: The following does not work correctly if there are any nulls in a parameterized array, because
-            // of null semantics (test: Contains_on_nullable_array_produces_correct_sql).
-            // https://github.com/aspnet/EntityFrameworkCore/issues/15892 tracks caching based on parameter values,
-            // which should allow us to enable this and have correct behavior.
-
-            // We still apply this translation when it's on a column expression, since that can't work anyway with
-            // EF Core's parameter to constant expansion
+            // Note: we exclude constant array expressions from this PG-specific optimization since the general
+            // EF Core mechanism is fine for that case. After https://github.com/aspnet/EntityFrameworkCore/issues/16375
+            // is done we may not need the check any more.
+            // Note: we exclude arrays/lists over Nullable<T> since the ADO layer doesn't handle them (but will in 5.0)
 
             if (method.IsClosedFormOf(Contains) &&
-                arrayOperand is ColumnExpression &&
-                //!(arrayOperand is SqlConstantExpression) &&   // When the null semantics issue is resolved
-                _sqlExpressionFactory.FindMapping(arrayOperand.Type) != null)
+                _sqlExpressionFactory.FindMapping(operand.Type) != null &&
+                !(operand is SqlConstantExpression) &&
+                !(operandElementType.IsGenericType && operandElementType.GetGenericTypeDefinition() == typeof(Nullable<>)))
```

Comment:
This could be replaces by `Nullable.GetUnderlyingType(operandElementType) is null`.